src/codegen/general.rs

```diff
@@ -16,32 +16,48 @@ use std::{
 
 pub fn start_comments(w: &mut dyn Write, conf: &Config) -> Result<()> {
     if conf.single_version_file.is_some() {
-        return start_comments_no_version(w);
+        start_comments_no_version(w, conf)
+    } else {
+        single_version_file(w, conf)
     }
-    writeln!(
-        w,
-        "// This file was generated by gir (https://github.com/gtk-rs/gir @ {})
-// from gir-files (https://github.com/gtk-rs/gir-files @ {})
-// DO NOT EDIT",
-        VERSION, conf.girs_version
-    )
 }
 
-pub fn start_comments_no_version(w: &mut dyn Write) -> Result<()> {
+pub fn start_comments_no_version(w: &mut dyn Write, conf: &Config) -> Result<()> {
     writeln!(
         w,
         "// This file was generated by gir (https://github.com/gtk-rs/gir)
-// from gir-files (https://github.com/gtk-rs/gir-files)
-// DO NOT EDIT"
+{}// DO NOT EDIT",
+        conf.girs_version
+            .iter()
+            .map(|info| {
+                format!(
+                    "// from {}{}\n",
+                    info.gir_dir.display(),
+                    info.get_repository_url()
+                        .map_or_else(String::new, |url| format!(" ({})", url)),
+                )
+            })
+            .collect::<String>()
     )
 }
 
 pub fn single_version_file(w: &mut dyn Write, conf: &Config) -> Result<()> {
-    writeln!(
+    write!(
         w,
         "Generated by gir (https://github.com/gtk-rs/gir @ {})
-from gir-files (https://github.com/gtk-rs/gir-files @ {})",
-        VERSION, conf.girs_version
+{}",
+        VERSION,
+        conf.girs_version
+            .iter()
+            .map(|info| {
+                format!(
+                    "from {} ({}@ {})\n",
```

Comment:
Or otherwise add it in the format string below. Looks like this now:

```
from /home/marijn/Code/aanmelder/gstreamer-rs/gir-files (https://github.com/gtk-rs/gir-files@ 21f7670)
from /home/marijn/Code/aanmelder/gstreamer-rs/gst-gir-files (@ ???)
```

Reply:
Weird, it shouldn't be the case:

```rust
info.get_repository_url()
    .map_or_else(String::new, |u| format!("{} ", u)),
```

Reply:
I think you pushed a fix for this right as I was posting this review. Let me re-check, should be good now.

Reply:
Yeah, looks good now:

```
+// Generated by gir (https://github.com/gtk-rs/gir @ 2a4d7f6+)
+// from gir-files (https://github.com/gtk-rs/gir-files @ 21f7670)
+// from gst-gir-files (@ ???)
```

Reply:
Yeay! \o/